File: packages/provide-uterm-go/hub/startup_broadcast_window_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

// Frames broadcast while a browser is still starting up must not be lost.
//
// A browser registers with deferBroadcast so its hello, hijack_state and
// presence_sync arrive before anything else; until it is activated it is not
// in the broadcast set at all, and what was broadcast meanwhile used to be
// dropped. That is right for frames the startup sequence already carries and
// wrong for the inspect channel, which has no replay: the browser builds that
// list from nothing, so a dropped http_req is a row missing for the rest of
// the session. Port of test_startup_broadcast_window.py.

import (
	"errors"
	"fmt"
	"strings"
	"testing"
)

func httpReqFrame(id, url string) map[string]any {
	return map[string]any{
		"type":     "http_req",
		"id":       id,
		"method":   "GET",
		"url":      url,
		"_channel": "http",
	}
}

// sentURLs returns the order the browser actually received request URLs in.
func sentURLs(ws *fakeBrowserWS, urls ...string) []string {
	var seen []string
	for _, payload := range ws.payloads() {
		for _, url := range urls {
			if strings.Contains(payload, url) {
				seen = append(seen, url)
			}
		}
	}
	return seen
}

func TestInspectFrameDuringStartupIsDeliveredOnActivation(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r1", "/api/users"))
	if got := len(ws.payloads()); got != 0 {
		t.Fatalf("a browser mid-startup must not be written to yet, got %d payload(s)", got)
	}

	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "/api/users"); len(got) != 1 {
		t.Fatalf("buffered inspect frame not delivered on activation, got %v", got)
	}
}

func TestBufferedInspectFramesKeepTheirOrder(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r0", "/api/zero"))
	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r1", "/api/one"))
	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r2", "/api/two"))
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	got := sentURLs(ws, "/api/zero", "/api/one", "/api/two")
	want := []string{"/api/zero", "/api/one", "/api/two"}
	if strings.Join(got, ",") != strings.Join(want, ",") {
		t.Fatalf("buffered frames reordered: got %v want %v", got, want)
	}
}

// TestSnapshotFromTheWindowIsDeliveredOnActivation: dropping snapshots assumed
// a newer one would follow. A terminal that emits one burst and goes idle
// produces exactly one, and losing it leaves the browser on the pre-burst
// screen its hello handed over.
func TestSnapshotFromTheWindowIsDeliveredOnActivation(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{"type": "snapshot", "screen": "ECHO_BANNER", "ts": 1.0})
	if got := len(ws.payloads()); got != 0 {
		t.Fatalf("a browser mid-startup must not be written to yet, got %d payload(s)", got)
	}

	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "ECHO_BANNER"); len(got) != 1 {
		t.Fatalf("buffered snapshot not delivered on activation, got %v", got)
	}
}

// TestOnlyTheNewestBufferedSnapshotSurvives pins the coalescing.
func TestOnlyTheNewestBufferedSnapshotSurvives(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	for index := range 5 {
		_ = h.Router.Broadcast(bg(), "w1", map[string]any{
			"type": "snapshot", "screen": fmt.Sprintf("screen-%d", index), "ts": float64(index),
		})
	}
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "screen-0", "screen-3", "screen-4"); len(got) != 1 || got[0] != "screen-4" {
		t.Fatalf("expected only the newest screen, got %v", got)
	}
}

func TestTerminalOutputFromTheWindowIsNotReplayed(t *testing.T) {
	// The hello's initial_snapshot already covers it; replaying prints twice.
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{"type": "term", "data": "ls -la\r\n"})
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := len(ws.payloads()); got != 0 {
		t.Fatalf("terminal output must not be replayed, got %d payload(s)", got)
	}
}

// TestPresenceSyncFromTheWindowIsDeliveredOnActivation pins the roster gap:
// the startup sequence sends each browser its own presence_sync, but that sync
// is computed at the browser's OWN join, so it cannot carry a user who arrives
// while the browser is still starting up. Dropping it left the roster one user
// short until some later presence event corrected it -- the failure behind
// TestDeckPresenceBroadcastSecondBrowser.
func TestPresenceSyncFromTheWindowIsDeliveredOnActivation(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{
		"type":   "presence_sync",
		"users":  []any{map[string]any{"user_id": "a"}, map[string]any{"user_id": "b"}},
		"config": map[string]any{},
	})
	if got := len(ws.payloads()); got != 0 {
		t.Fatalf("a browser mid-startup must not be written to yet, got %d payload(s)", got)
	}

	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "presence_sync"); len(got) != 1 {
		t.Fatalf("buffered presence_sync not delivered on activation, got %v", got)
	}
}

// TestPresenceLeaveFromTheWindowIsDeliveredOnActivation: worse than a missed
// sync, because a leave is a delta -- dropping it keeps a ghost user in the
// list with nothing to reconcile it against.
func TestPresenceLeaveFromTheWindowIsDeliveredOnActivation(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{"type": "presence_leave", "user_id": "departed"})
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "departed"); len(got) != 1 {
		t.Fatalf("buffered presence_leave not delivered on activation, got %v", got)
	}
}

// TestControlTransferFromTheWindowIsDeliveredOnActivation: who is driving is a
// delta too. The startup presence_sync stamps is_owner as of this browser's
// join, and nothing restates a handover that happens inside the window.
func TestControlTransferFromTheWindowIsDeliveredOnActivation(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{
		"type": "control_transfer", "from_user_id": "a", "to_user_id": "b", "reason": "handover",
	})
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := sentURLs(ws, "control_transfer"); len(got) != 1 {
		t.Fatalf("buffered control_transfer not delivered on activation, got %v", got)
	}
}

// TestPresenceUpdateFromTheWindowIsNotReplayed pins the deliberate exclusion:
// transient per-user state that the next update supersedes, frequent enough to
// crowd out the buffer's cap.
func TestPresenceUpdateFromTheWindowIsNotReplayed(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	_ = h.Router.Broadcast(bg(), "w1", map[string]any{
		"type": "presence_update", "user_id": "a", "name": "A", "color": "#fff", "role": "viewer",
	})
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	if got := len(ws.payloads()); got != 0 {
		t.Fatalf("presence_update must stay dropped, got %d payload(s)", got)
	}
}

func TestActivatedBrowserReceivesInspectFramesDirectly(t *testing.T) {
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)
	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r1", "/api/users"))

	if got := sentURLs(ws, "/api/users"); len(got) != 1 {
		t.Fatalf("activated browser missed a live inspect frame, got %v", got)
	}
	h.lock.Lock()
	_, buffered := h.startupPendingFrames[ws]
	h.lock.Unlock()
	if buffered {
		t.Fatal("an activated browser must not accumulate a backlog")
	}
}

func TestStartupBufferIsCappedRatherThanUnbounded(t *testing.T) {
	// A browser that never activates must not be able to grow this forever.
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)

	for i := 0; i < startupBufferMaxFrames+25; i++ {
		_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r", "/api/x"))
	}

	h.lock.Lock()
	queued := len(h.startupPendingFrames[ws])
	h.lock.Unlock()
	if queued != startupBufferMaxFrames {
		t.Fatalf("buffer not capped: got %d want %d", queued, startupBufferMaxFrames)
	}
}

func TestDisconnectingBrowserDropsItsBacklog(t *testing.T) {
	// Nothing will ever flush it, so holding it is a leak.
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)
	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r1", "/api/users"))

	_, _ = h.Conn.CleanupBrowserDisconnect(bg(), "w1", ws, false)

	h.lock.Lock()
	_, buffered := h.startupPendingFrames[ws]
	h.lock.Unlock()
	if buffered {
		t.Fatal("a disconnected browser kept its backlog")
	}
}

func TestSocketThatCannotTakeItsBacklogDropsIt(t *testing.T) {
	// Pending is the right resting state for a socket that just failed a
	// write: the broadcast path skips it rather than retrying into a dead
	// connection, and the disconnect path clears both.
	h, _ := newTestHub(t, nil)
	ws := newBrowserWS("b")
	_, _ = h.Conn.RegisterBrowser(bg(), "w1", ws, "viewer", true)
	_ = h.Router.Broadcast(bg(), "w1", httpReqFrame("r1", "/api/users"))
	ws.failSend = errors.New("socket gone")

	h.Conn.ActivateBrowserBroadcasts(bg(), "w1", ws)

	h.lock.Lock()
	_, buffered := h.startupPendingFrames[ws]
	stillPending := h.startupPendingBrowsers[ws]
	h.lock.Unlock()
	if buffered {
		t.Fatal("a failed flush must drop the backlog")
	}
	if !stillPending {
		t.Fatal("a socket that failed a write must stay skipped")
	}
}

File: packages/provide-uterm-ts/src/hub/startup-broadcast-window.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Frames broadcast while a browser is still starting up must not be lost.
 *
 * A browser registers with deferBroadcast so its hello, hijack_state and
 * presence_sync arrive before anything else; until it is activated it is not in
 * the broadcast set at all, and what was broadcast meanwhile used to be dropped
 * outright — this port had the pending set but no buffer behind it.
 *
 * Port of test_startup_broadcast_window.py.
 */

import { describe, expect, it } from "vitest";
import {
  type ConnectionHubCallbacks,
  ConnectionManager,
  MessageRouter,
  type RouterHubCallbacks,
  STARTUP_BUFFER_MAX_FRAMES,
  WorkerRegistry,
  type WorkerTermState,
} from "./index.ts";
import type { Connection } from "./models.ts";

const WORKER = "w1";

function httpReq(id: string, url: string): Record<string, unknown> {
  return { type: "http_req", id, method: "GET", url, _channel: "http" };
}

/** A browser socket that records what it was sent, or refuses to take it. */
class FakeBrowser {
  readonly sent: string[] = [];
  #throws = false;

  failWrites(): void {
    this.#throws = true;
  }

  async sendText(payload: string): Promise<void> {
    if (this.#throws) {
      throw new Error("socket closed");
    }
    this.sent.push(payload);
  }
}

/** One fake standing in for both hub surfaces, so they share the same state. */
class FakeHub implements RouterHubCallbacks, ConnectionHubCallbacks {
  readonly registry = new WorkerRegistry<WorkerTermState>();
  readonly startupPendingBrowsers = new Set<Connection>();
  readonly startupPendingFrames = new Map<Connection, Record<string, unknown>[]>();
  maxWorkers = 100;
  maxConnectionsPerPrincipal = 100;

  isHijacked(): boolean {
    return false;
  }

  isDashboardHijackActive(): boolean {
    return false;
  }

  hasValidRestLease(): boolean {
    return false;
  }

  async removeDeadBrowsers(): Promise<boolean> {
    return false;
  }
}

/** A hub with one worker and one browser held mid-handshake. */
async function pending() {
  const hub = new FakeHub();
  const router = new MessageRouter({ hub, sendTimeoutS: 0.05 });
  const connections = new ConnectionManager({ hub, now: () => 0 });
  const browser = new FakeBrowser();
  await connections.registerBrowser(WORKER, browser, "viewer", { deferBroadcast: true });
  return { hub, router, connections, browser };
}

/** Which of `needles` the browser was actually sent, in order. */
function seen(browser: FakeBrowser, ...needles: string[]): string[] {
  return browser.sent.flatMap((payload) => needles.filter((needle) => payload.includes(needle)));
}

describe("the broadcast window a connecting browser cannot see", () => {
  it("delivers an inspect frame sent during startup on activation", async () => {
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, httpReq("r1", "/api/users"));
    expect(browser.sent).toEqual([]);

    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(seen(browser, "/api/users")).toEqual(["/api/users"]);
  });

  it("keeps buffered inspect frames in arrival order", async () => {
    const { router, connections, browser } = await pending();

    for (const index of [0, 1, 2]) {
      await router.broadcast(WORKER, httpReq(`r${index}`, `/api/${index}`));
    }
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(seen(browser, "/api/0", "/api/1", "/api/2")).toEqual(["/api/0", "/api/1", "/api/2"]);
  });

  it("does not replay terminal output from the window", async () => {
    // The hello's initial_snapshot already covers it; replaying prints twice.
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, { type: "term", data: "ls -la\r\n", ts: 1 });
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(browser.sent).toEqual([]);
  });

  it("delivers a presence_sync from the window on activation", async () => {
    // The startup presence_sync is computed at THIS browser's join, so it
    // cannot carry a user who arrives while it is still starting up.
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, {
      type: "presence_sync",
      users: [{ user_id: "a" }, { user_id: "b" }],
      config: {},
    });
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(seen(browser, "presence_sync")).toEqual(["presence_sync"]);
  });

  it("delivers a presence_leave from the window on activation", async () => {
    // Worse than a missed sync: a delta, so dropping it keeps a ghost user.
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, { type: "presence_leave", user_id: "departed" });
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(seen(browser, "departed")).toEqual(["departed"]);
  });

  it("delivers a control_transfer from the window on activation", async () => {
    // Who is driving is a delta too, and nothing restates it.
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, {
      type: "control_transfer",
      from_user_id: "a",
      to_user_id: "b",
      reason: "handover",
    });
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(seen(browser, "control_transfer")).toEqual(["control_transfer"]);
  });

  it("does not replay a presence_update from the window", async () => {
    // Transient per-user state the next update supersedes.
    const { router, connections, browser } = await pending();

    await router.broadcast(WORKER, {
      type: "presence_update",
      user_id: "a",
      name: "A",
      color: "#fff",
      role: "viewer",
    });
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(browser.sent).toEqual([]);
  });

  it("sends inspect frames straight to an activated browser", async () => {
    const { router, connections, browser } = await pending();
    await connections.activateBrowserBroadcasts(WORKER, browser);

    await router.broadcast(WORKER, httpReq("r1", "/api/users"));

    expect(seen(browser, "/api/users")).toEqual(["/api/users"]);
  });

  it("caps the buffer rather than growing it without bound", async () => {
    // A browser that never activates must not be able to grow this forever.
    const { hub, router, browser } = await pending();

    for (let index = 0; index < STARTUP_BUFFER_MAX_FRAMES + 10; index += 1) {
      await router.broadcast(WORKER, httpReq(`r${index}`, `/api/${index}`));
    }

    expect(hub.startupPendingFrames.get(browser)?.length).toBe(STARTUP_BUFFER_MAX_FRAMES);
  });

  it("drops the backlog when the browser disconnects", async () => {
    const { hub, router, connections, browser } = await pending();

    await router.broadcast(WORKER, httpReq("r1", "/api/users"));
    connections.cleanupBrowserDisconnect(WORKER, browser, false);

    expect(hub.startupPendingFrames.has(browser)).toBe(false);
    expect(hub.startupPendingBrowsers.has(browser)).toBe(false);
  });

  it("leaves a socket that cannot take its backlog pending", async () => {
    // Pending means the broadcast path skips it, which is the right resting
    // state for a connection that just failed a write.
    const { hub, router, connections, browser } = await pending();

    await router.broadcast(WORKER, httpReq("r1", "/api/users"));
    browser.failWrites();
    await connections.activateBrowserBroadcasts(WORKER, browser);

    expect(hub.startupPendingBrowsers.has(browser)).toBe(true);
    expect(hub.startupPendingFrames.has(browser)).toBe(false);
  });
});
